pymysqlpool/pool.py

```diff
@@ -141,23 +145,43 @@ def _wait(self):
 
     def get_conn(self):
         with self.cond:
-            # Lack of resources and wait
-            if len(self.unuse_list) <= 0 and \
-                    self.current_size >= self.max_size:
-                # note: TimeoutError mean release operation exception
-                # or max_size much less than concurrence
-                self.cond.wait_for(self._wait, self.timeout)
-                if len(self.unuse_list) <= 0:
-                    raise TimeoutError
-            # Lack of resources but can created
-            if len(self.unuse_list) <= 0 and \
-                    self.current_size < self.max_size:
-                self.create_conn()
-
-            self.current_size += 1
-            c = self.unuse_list.pop()
-            self.inuse_list.add(c)
-            return c
+            return self.__get_conn()
+
+    def __get_conn(self, retry_count=0):
+        # Lack of resources and wait
+        if len(self.unuse_list) <= 0 and \
+                self.current_size >= self.max_size:
+            # note: TimeoutError mean release operation exception
+            # or max_size much less than concurrence
+            self.cond.wait_for(self._wait, self.timeout)
+            if len(self.unuse_list) <= 0:
+                raise TimeoutError
+        # Lack of resources but can created
+        if len(self.unuse_list) <= 0 and \
+                self.current_size < self.max_size:
+            self.create_conn()
+
+        return self.__get_safe_conn(retry_count)
+
+    def __get_safe_conn(self, retry_count):
+        self.current_size += 1
+        c = self.unuse_list.pop()
+        if self.ping_check:
+            now = int(time())
+            timeout = now
+            if isinstance(int, self.ping_check):
+                timeout = timeout - self.ping_check
+            if not hasattr(c, '__ping_check_timestamp'):
+                c.__ping_check_timestamp = now
+            try:
+                if c.__ping_check_timestamp < timeout:
+                    c.__ping_check_timestamp = now
+                    c.ping()
+            except:
+                self.current_size -= 1
+                if retry_count < 10: c = self.__get_conn(retry_count+1)
```

Comment:
Assuming that retry exceeds 10, ping() still fails. is there need to remove this invalid conn and recreate a conn, or throw an exception?

Reply:
When the `ping()` fails, the _conn_ is discarded and another _conn_ is fetched from the pool. The _retry_count_ is used to to get and return the first healthy _conn_ from the pool a couple of times. If the _conn_ object fails on trying reconnect, maybe the object has some problem with its internal properties. But... if other instances have the same problem, maybe there is a network problem on going, and no _conn_ will be returned anymore. This implementation just return a `None` _conn_ on a network problem occurring, but we can make a custom exception if you prefer.

Reply:
Oh, yes, `ping()` will [reconnect](https://pymysql.readthedocs.io/en/latest/modules/connections.html?highlight=ping#pymysql.connections.Connection.ping), I forgot it 😅...

Nice pr 👍🎉👏, merged.